### DISCUSSION TITLE
How to get element's calculation attributes ?

### DISCUSSION THREAD

**Discussion by Lefort-Antoine** [cwapi3d/cwapi3dpython#208] [Q&A]
I'm a beginner.
I found no way to get all the attribute of an element. Not only the auto or specific as beam list I need all of them ! 
Thanks for your help

**Comment by Brunner246** [comment]
Hi,

If you know the range, you can iterate over the elements and store the attributes in a list.

```python
element_attributes_map = defaultdict(list)
your_attribute_range = range(1, 80) # check for attributes from userattribute 1 to userattribute 80

for element_id in element_ids:
    for value in your_attribute_range:
        attribute_value = ac.get_user_attribute(element_id, value)
        if attribute_value:
            element_attributes_map[element_id].append(attribute_value)

            
for key, value in element_attributes_map.items():
    print(f"Element {key} : attributes {value}")

```

same, but more compact
```python
element_attributes_map = {
    element_id: [ac.get_user_attribute(element_id, value) for value in your_attribute_range if ac.get_user_attribute(element_id, value)]
    for element_id in element_ids
}
```

**Reply by Frmnt-Julien** [reply]
Hi,

Thanks for your feedback, I am a collaborator of Antoine, I use cadwork every day but I am a beginner in programming.

I tested the solution given but it only gives the information of user attributes from 1 to 100 and more. How to have the other attributes? more precisely the auto-attributes (example: Total prod. list).

Thanks in advance

![attributs](https://github.com/user-attachments/assets/c7c07130-6013-4a35-b0c7-686e018fc0ad)

**Comment by Brunner246** [comment]
To get all used "user attributes" you can use the function `attribute_controller.get_user_attribute_count()`

To get the total production list count, you have to setup a datastructure by yourself. 

```python
# cadwork imports
from collections import defaultdict

element_same_production_nr = defaultdict(list)

for element in active_identifiable_element_ids:
    production_nr = ac.get_production_number(element)
    element_same_production_nr[production_nr].append(element)
    
for production_nr, elements in element_same_production_nr.items():
    print(f"Production Number: {production_nr}")
    for element in elements:
        print(f"Element ID: {element}")
````
